x-pack/plugins/lens/public/editor_frame_service/editor_frame/config_panel/layer_panel.test.tsx

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { act } from 'react-dom/test-utils';
+import {
+  createMockVisualization,
+  createMockFramePublicAPI,
+  createMockDatasource,
+  DatasourceMock,
+} from '../../mocks';
+import { EuiFormRow, EuiPopover } from '@elastic/eui';
+import { mount } from 'enzyme';
+import { mountWithIntl } from 'test_utils/enzyme_helpers';
+import { Visualization } from '../../../types';
+import { LayerPanel } from './layer_panel';
+import { coreMock } from 'src/core/public/mocks';
+import { generateId } from '../../../id_generator';
+
+jest.mock('../../../id_generator');
+
+describe('LayerPanel', () => {
+  let mockVisualization: jest.Mocked<Visualization>;
+  let mockDatasource: DatasourceMock;
+
+  function getDefaultProps() {
+    const frame = createMockFramePublicAPI();
+    frame.datasourceLayers = {
+      first: mockDatasource.publicAPIMock,
+    };
+    return {
+      layerId: 'first',
+      activeVisualizationId: 'vis1',
+      visualizationMap: {
+        vis1: mockVisualization,
+      },
+      activeDatasourceId: 'ds1',
+      datasourceMap: {
+        ds1: mockDatasource,
+      },
+      datasourceStates: {
+        ds1: {
+          isLoading: false,
+          state: 'state',
+        },
+      },
+      visualizationState: 'state',
+      updateVisualization: jest.fn(),
+      updateDatasource: jest.fn(),
+      updateAll: jest.fn(),
+      framePublicAPI: frame,
+      isOnlyLayer: true,
+      onRemoveLayer: jest.fn(),
+      dispatch: jest.fn(),
+      core: coreMock.createStart(),
+    };
+  }
+
+  beforeEach(() => {
+    mockVisualization = {
+      ...createMockVisualization(),
+      id: 'testVis',
+      visualizationTypes: [
+        {
+          icon: 'empty',
+          id: 'testVis',
+          label: 'TEST1',
+        },
+      ],
+    };
+
+    mockVisualization.getLayerIds.mockReturnValue(['first']);
+    mockDatasource = createMockDatasource('ds1');
+  });
+
+  it('should fail to render if the public API is out of date', () => {
```

Comment:
I don't understand this test. Could you explain? 🙏

Reply:
It's testing another one of the edge cases in rendering, where the known layers don't match. This tests that when rendering layer `first` without any information about that ID, it doesn't throw errors.